C: Replace `tb_client.h` with script that generates it

kprotty opened the pull request:
`tb_client.h` gets out of sync so much that it's better if we auto generate it for the near future. TigerBeetle clients who build `tb_client` also get the generated header in `zig-out` and it's no longer tracked in git versioning to avoid manual sync issues.

Commit:
C: Replace tb_client.h which script that generates it

`tb_client.h` gets out of sync so much that it's better if we auto generate it for the near future. TigerBeetle clients who build `tb_client` also get the generated header in `zig-out` and it's no longer tracked in git versioning to avoid manual sync issues.

sentientwaffle commented on build.zig:
```diff
@@ -78,11 +78,20 @@ pub fn build(b: *std.build.Builder) void {
     }
 
     {
+        const output_dir = "zig-out";
```
Zig's `LibExeObjStep` has a built-in `output_dir` property, can you use that instead?

kprotty replied:
Unfortunately, it's not set during `fn build()`.

sentientwaffle commented on src/c/tb_client_header.zig:
```diff
@@ -0,0 +1,217 @@
+}
+
+pub fn main() !void {
+    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
```
```suggestion
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
```

sentientwaffle commented on src/c/tb_client_header.zig:
```diff
@@ -0,0 +1,217 @@
+const std = @import("std");
+const tb = @import("../tigerbeetle.zig");
+const build_options = @import("build_options");
```
```suggestion
const build_options = @import("build_options");

const tb = @import("../tigerbeetle.zig");
```

sentientwaffle commented on src/c/tb_client_header.zig:
```diff
@@ -0,0 +1,217 @@
+const tb = @import("../tigerbeetle.zig");
+const build_options = @import("build_options");
+
+fn c_type_name(comptime ty: type) []const u8 {
```
Maybe call this `Type` instead of `ty`? I think zig mostly has the convention of using upper-case for types, even when they are being used as variables like this.

sentientwaffle reviewed (commented):
Can you rebase this onto main?

kprotty commented:
Should we leave `src/c/tb_client.h` in source and update it with that generated by `zig build tb_client`? Or should we remove it to avoid syncing? @sentientwaffle

sentientwaffle commented:
Since it is a source file (albeit generated) and not too large, let's leave it checked in. This way we can easily see its changes over time in git history — and it is always obvious when it is modified.

We should add a github workflow to verify that it is clean (i.e. does not need to be regenerated).

batiati commented:
> Should we leave `src/c/tb_client.h` in source and update it with that generated by `zig build tb_client`? Or should we remove it to avoid syncing? @sentientwaffle

It's worth noting that [src/c/test.zig](https://github.com/tigerbeetledb/tigerbeetle/blob/7fbac6dc9d5e438086e7d131f46a86d8460dc29b/src/c/test.zig#L6) uses a `@cImport(@cInclude("tb_client.h"))` to test the FFI API.

sentientwaffle commented:
Thanks, @batiati. So in `build.zig` the `zig build test` step should depend on the `tb_client.h` generation step, to make sure those are synced up.

Commit:
build: dependOn generating tb_client.h for unit_test & libtb_client

The unit test runs a script which reads tb_client.h with cImport and verifies all the values match the zig ones. To avoid the header going out of sync on test, this re-generates the header.

The `tb_client_header.zig` generator script also now writes directly to source (`src/c/tb_client.h`).

Commit:
client: Cleanup + autogen more in tb_client.h

- Generate packet/status/operation typedefs
- Give C enum variants concrete values instead of relying on definition order
- Replace C enum `TB_OP_` prefix with `TB_OPERATION_`

Commit:
client: zig fmt + update Go header & TB_OP references

TB_OP prefix was renamed to TB_OPERATION

kprotty commented:
I think the `@compileError` in `src/c/tb_client.zig:global_allocator` is getting eagerly evaluated when it shouldn't be. You'd normally fix this by hiding it behind a decl: 
```zig
// else @compileError();
else no_allocator;
const no_allocator = @compileError();
```

Is there anywhere else that we do this pattern?

Commit:
go: fix C header enum and error references

Commit:
go: move compileError for lack of libc outside allocator decl

Zig evaluates the if statement eagerly for `global_allocator`. This can trigger the `@compileError` even when the branch is not active. Moving the error out to its own declaration prevents it from being errornously/immediately evaluated.

sentientwaffle commented on build.zig:
```diff
@@ -45,6 +45,47 @@ pub fn build(b: *std.build.Builder) void {
+
+        const step = b.step("benchmark_segmented_array", "Benchmark SegmentedArray search");
+        step.dependOn(&run_cmd.step);
+    }
```
These shouldn't be here; probably artifacts from your rebase.

kprotty replied:
Yea, after requested for review I thought the same thing.

sentientwaffle commented on build.zig:
```diff
@@ -45,6 +45,47 @@ pub fn build(b: *std.build.Builder) void {
+    }
+
+    // Executable which generates src/c/tb_client.h
+    const tb_client_header_generate = blk: {
```
I think "generate" is redundant (in relation to "build") since it will be run `zig build tb_client_header`.

kprotty replied:
`zig build tb_client_header` would be a thing only if we have a `b.addStep()` for it. ATM, it's only an internal step that's a dependency for `zig build test` and `zig build tb_client`. Doesn't have to be for the latter, but updating the header in source when modifying tb_cilent serves as a reminder to update it for the clients too (maybe automate this as well since all clients are in tree?)

sentientwaffle replied:
Ah, good point. Yes, we should automate updating the clients too, but you can open a github issue for that to defer it to a future PR.

sentientwaffle commented on src/c/tb_client_header.zig:
```diff
@@ -0,0 +1,213 @@
+        char.* = input[i];
+        char.* -= 32 * @as(u8, @boolToInt(char.* >= 'a' and char.* <= 'z'));
+    }
+    return &output;
```
Interesting! I haven't seen this `comptime` pattern before.

kprotty replied:
Do you refer to returning the local output array or the branchless programming technique?

sentientwaffle replied:
I was referring to declaring an array and then returning a reference/slice to it from within the function that declares it. In a non-comptime setting that wouldn't be legal (since the array is on the stack). But in comptime I guess it works!

Commit:
c: move allocator compileError into runtime panic

`tb_client_header.zig` seemed to be eagerly evaluating the global_allocator declaration when it does `@import("tb_client.zig")` for generating the header. This triggered the `@compileError` if the header generation script was running on linux. Running `zig build tb_client` has the header gen as a dependency so that build commands fails on linux.

This moves the `@compileError` into a panic at runtime so `tb_client_header.zig` can inspect `tb_client.zig` for header generation without error-ing out at compile time.

Commit:
c/client: bump node version + zig fmt

sentientwaffle commented on src/c/tb_client.zig:
```diff
@@ -63,7 +64,7 @@ else if (builtin.target.os.tag == .windows)
     }).gpa.allocator()
 else
-    @compileError("tb_client must be built with libc");
+    null;
```
What was the reason for this change? Can we not detect a missing allocator at comptime?

kprotty replied:
It's a bit hidden in the commit message now, but essentially `tb_client_header` would run (on linux, without libc) to generate the tb_client.h. And to do that, it would `@import("tb_client.zig")` for comptime inspection of its types. This may have eagerly evaluated `const global_allocator =` (as it needed to know the type of that decl) which caused the `@compileError` to be hit unexpectedly. 

This is my understanding of the problem at least.

Hiding the `@compileError` behind a decl to imply less strict evaluated didn't seem to help. In build.zig, I then tried adding `.linkLibC()` to the `tb_client_header` executable step, but this resulting in that glibc error I spoke of earlier today:
```bash
error(compilation): tigerbeetle/zig/lib/libc/glibc/sysdeps/x86_64/start-2.33.S:1:1: unable to build C object: FileNotFound
error(compilation): tigerbeetle/zig/lib/libc/glibc/csu/elf-init-2.33.c:1:1: unable to build C object: FileNotFound
error: unable to build glibc CRT file: BuildingLibCObjectFailed
```

sentientwaffle replied:
What if `global_allocator` was declared+defined in `init()`? (Or what is the reason for it to be "global"?)

kprotty replied:
I thought that would work, butI finally realized that `tb_client_init_{echo}` being marked as `export fn` was the issue as it forced their functions (which used `global_allocator`) to be compiled/codegen'd into the `tb_client_header.zig` executable even if they weren't referenced. 

Removing the `export` basically fixes it, but im thinking of how to [conditionally export](https://ziglang.org/documentation/0.10.0/#export) it based on if it's apart of `tb_client_header` or `tb_client`.

sentientwaffle commented on src/c/tb_client_header.zig:
```diff
@@ -0,0 +1,213 @@
+    switch (@typeInfo(Type)) {
+        .Array => |info| return resolve_c_type(info.child),
+        .Enum => |info| return resolve_c_type(info.tag_type),
+        .Struct => return resolve_c_type(std.meta.Int(.unsigned, @sizeOf(Type) * 8)),
```
Can we use `@bitSizeOf()` here?

sentientwaffle reviewed (commented):
This looks great, @kprotty! Just a few notes/questions

Commit:
c: Only export tb_client functions when building shared library

It seems the real cause of the `global_allocator` getting erronously evaluated by `@import()` is due to the tb_cilent functions being declared as `export`. This forced them to be codegen (hence evaluated) even when not building as the shared library.

This moves `global_allocator` to its only use-site for clarity and, more importantly, conditionally exports the tb_client functions only if the zig file is being used to compile the shared library (and not being imported)

Commit:
c: use bitSizeOf over sizeOf * 8 + zig fmt

Commit:
c: Export tb_client functions if libc + tb_client links libc on windows

`c/test.zig` relies on `cImport()` to access tb_client functions. This is good for testing as it ensures the C header matches zig. The C functions were only being exported for `zig build tb_client` and not `zig build test`, so now they're exported anything libc is being linked (which is both cases).

Commit:
c: Add comment to C header that it was auto-generated.

Just an extra warning to those contributing on the C client.
